extensions/scarb-snforge-test-collector/src/main.rs

```diff
@@ -0,0 +1,47 @@
+use anyhow::Result;
+use camino::Utf8PathBuf;
+use clap::Parser;
+use std::str::FromStr;
+
+use scarb_metadata::MetadataCommand;
+
+use crate::compilation::compile_tests;
+use crate::crate_collection::collect_test_compilation_targets;
+use crate::metadata::{
+    corelib_for_package, dependencies_for_package, id_for_package, paths_for_package,
+};
+
+mod compilation;
+mod crate_collection;
+mod metadata;
+
+/// Starknet Foundry private extension for compiling test artifacts.
+/// Users should not call it directly.
+#[derive(Parser)]
+#[command(version)]
+struct Args {
+    #[arg(long)]
+    output_path: Utf8PathBuf,
+}
+
+fn main() -> Result<()> {
+    let args = Args::parse();
+    let scarb_metadata = MetadataCommand::new().inherit_stderr().exec()?;
+
+    let manifest_path = Utf8PathBuf::from_str(&std::env::var("SCARB_MANIFEST_PATH")?)?;
+    let package_id = id_for_package(&scarb_metadata, &manifest_path)?;
+
+    let (package_path, package_source_dir_path, package_name) =
+        paths_for_package(&scarb_metadata, &package_id)?;
+    let dependencies = dependencies_for_package(&scarb_metadata, &package_id)?;
+    let corelib_path = corelib_for_package(&scarb_metadata, &package_id)?;
+
+    let compilation_targets =
+        collect_test_compilation_targets(&package_path, &package_name, &package_source_dir_path)?;
+    let test_crates = compile_tests(&compilation_targets, &dependencies, &corelib_path)?;
+
+    let serialized_crates = serde_json::to_string(&test_crates)?;
+    std::fs::write(args.output_path, serialized_crates)?;
```

Comment:
```suggestion
    let file = std::fs::File::create(args.output_path.clone())?;
    let file = BufWriter::new(&file);
    serde_json::to_writer(file, &test_crates)
        .context("failed to serialize test crates")?;
```

Reply:
Will foundry make sure that the containing directory of output_path exists? Both `std::fs` functions declare `Depending on the platform, this function may fail if the full directory path does not exist`

Reply:
Shouldn't we lock here? cc'ing @mkaput as our lock expert 👑

Reply:
Yup, it will be created by `snforge`. I thought about `workspace_root/snforge_target/{package_name}.sierra.json` as the output path. I chose a separate directory to make sure Scarb doesn't overwrite anything in `target` directory

Reply:
Having a custom `snforge_target` directory, you basically introduce a new concept that users will have to put in their `gitignore` and also it will hurt their eyes IMHO.

I'd definitely keep this in the `target/<profile name>` hierarchy. You can freely use the `target/<profile name>/snforge` directory.

That said, it'd be safer to use our file locking machinery here, for safety purposes. Who knows, when some user will attempt to run `snforge build` twice 😅 For this, you have to use our [`Filesystem`](https://github.com/software-mansion/scarb/blob/b78262ec74991546bc8c3d2383abb2b6f57f5260/scarb/src/flock.rs#L130) data structure and its [`open_rw`](https://github.com/software-mansion/scarb/blob/b78262ec74991546bc8c3d2383abb2b6f57f5260/scarb/src/flock.rs#L187) method. Here you have a snippet from our compilers that writes JSON to such files:

https://github.com/software-mansion/scarb/blob/a5d8783c9936db66be138704ffdcf7e01457ff02/scarb/src/compiler/helpers.rs#L46-L58

An extra benefit is that `scarb clean` will handle it for free for you.